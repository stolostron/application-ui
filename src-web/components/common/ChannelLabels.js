// Copyright (c) 2020 Red Hat, Inc.
'use strict'

import React from 'react'
import { Icon } from 'carbon-components-react'
import PropTypes from 'prop-types'
import {
  Divider,
  Split,
  SplitItem,
  Stack,
  StackItem
} from '@patternfly/react-core'
import R from 'ramda'
import LabelWithPopover from './LabelWithPopover'
import { getSearchLink } from '../../../lib/client/resource-helper'
import msgs from '../../../nls/platform.properties'
import resources from '../../../lib/shared/resources'

resources(() => {
  require('../../../scss/channel-labels.scss')
})
const ChannelLabels = ({ channels, locale, showSubscriptionAttributes=true }) => {
  const groupByChannelType = R.groupBy(ch => {
    const channelType = (ch.type && ch.type.toLowerCase()) || ''
    return channelType === 'github' ? 'git' : channelType
  })
  const channelMap = groupByChannelType(channels || [])
  // Create sorting function for channels
  const channelSort = R.sortWith([
    R.ascend(R.prop('pathname')),
    R.ascend(R.prop('gitBranch')),
    R.ascend(R.prop('gitPath'))
  ])
  return (
    <div className="label-with-popover-container channel-labels">
      {['git', 'helmrepo', 'namespace', 'objectbucket']
        .filter(chType => channelMap[chType])
        .map(chType => {
          return (
            <LabelWithPopover
              key={`${chType}`}
              labelContent={
                <Split hasGutter className="channel-label">
                  <SplitItem>
                    {msgs.get(`channel.type.${chType}`, locale)}
                    {channelMap[chType].length > 1 &&
                      ` (${channelMap[chType].length})`}
                  </SplitItem>
                  <SplitItem>
                    <Icon className="channel-entry-icon" name="icon--launch" />
                  </SplitItem>
                </Split>
              }
            >
              <Stack className="channel-labels channel-labels-popover-content">
                {channelSort(channelMap[chType]).map((channel, index) => {
                  const pathname = channel.pathname
                  const link =
                    chType === 'namespace'
                      ? getSearchLink({
                        properties: {
                          namespace: pathname,
                          kind: 'deployable',
                          apigroup: 'apps.open-cluster-management.io'
                        }
                      })
                      : pathname
                  return (
                    <React.Fragment
                      key={`${chType}-${channel.pathname}-${
                        channel.gitBranch
                      }-${channel.gitPath}`}
                    >
                      {index > 0 && (
                        <StackItem>
                          <Divider />
                        </StackItem>
                      )}
                      <StackItem className="channel-entry">
                        <Stack hasGutter>
                          <StackItem className="channel-entry-link">
                            <a href={link} target="_blank">
                              <Split hasGutter>
                                <SplitItem>
                                  <Icon
                                    className="channel-entry-icon"
                                    name="icon--launch"
                                  />
                                </SplitItem>
                                <SplitItem>{pathname}</SplitItem>
                              </Split>
                            </a>
                          </StackItem>
                          {showSubscriptionAttributes && chType === 'git' && (
                            <React.Fragment>
                              {['gitBranch', 'gitPath'].map(attrib => {
                                return (
                                  <StackItem
                                    key={attrib}
                                    className="channel-entry-attribute"
                                  >
                                    <Split hasGutter>
                                      <SplitItem className="channel-entry-attribute-name">
                                        {msgs.get(
                                          `channel.type.label.${attrib}`
                                        )}
                                      </SplitItem>
                                      <SplitItem>
                                        {channel[attrib]
                                          ? channel[attrib]
                                          : msgs.get(
                                            'channel.type.label.noData',
                                            locale
                                          )}
                                      </SplitItem>
                                    </Split>
                                  </StackItem>
                                )
                              })}
                            </React.Fragment>
                          )}
                        </Stack>
                      </StackItem>
                    </React.Fragment>
                  )
                })}
              </Stack>
            </LabelWithPopover>
          )
        })}
    </div>
  )
}

ChannelLabels.propTypes = {
  channels: PropTypes.arrayOf(PropTypes.object),
  locale: PropTypes.string,
  showSubscriptionAttributes: PropTypes.bool
}

export default ChannelLabels
