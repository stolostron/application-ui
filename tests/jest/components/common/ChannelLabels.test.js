/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
"use strict";

import React from "react";
import { Icon } from "carbon-components-react";
import ChannelLabels from "../../../../src-web/components/common/ChannelLabels";
import renderer from "react-test-renderer";

describe("ChannelLabels", () => {
  it("renders as expected", () => {
    const component = renderer.create(
      <ChannelLabels
        channels={[
          { type: "GitHub", pathname: "https://github.com/org/repo.git" },
          { type: "git", pathname: "https://github.com/org/repo2.git" },
          { type: "namespace", pathname: "sample-ns" }
        ]}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
