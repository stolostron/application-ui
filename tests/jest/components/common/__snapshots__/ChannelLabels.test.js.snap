// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelLabels renders as expected 1`] = `
<div
  className="label-with-popover-container channel-labels"
>
  <div
    className="label-with-popover"
  >
    <span
      className="pf-c-label"
      onClick={[Function]}
      onPointerEnter={[Function]}
      onPointerLeave={[Function]}
    >
      <a
        className="pf-c-label__content"
        href="#"
      >
        <div
          className="pf-l-split pf-m-gutter channel-type-label"
        >
          <div
            className="pf-l-split__item"
          >
            Git
          </div>
          <div
            className="pf-l-split__item"
          >
            <svg
              alt="Provide a description that will be used as the title"
              aria-label="Provide a description that will be used as the title"
              className="channel-type-icon"
              fillRule="evenodd"
              height="16"
              name="icon--launch"
              role="img"
              viewBox="0 0 16 16"
              width="16"
            >
              <title>
                Provide a description that will be used as the title
              </title>
              <path
                d="M14.3 1h-3.8V0H16v5.5h-1V1.7L9.7 7 9 6.3 14.3 1z"
              />
              <path
                d="M14.3 1h-3.8V0H16v5.5h-1V1.7L9.7 7 9 6.3 14.3 1z"
              />
              <path
                d="M13 9h1v6c0 .6-.4 1-1 1H1c-.6 0-1-.4-1-1V3c0-.6.4-1 1-1h7v1H1v12h12V9z"
              />
            </svg>
          </div>
        </div>
      </a>
    </span>
  </div>
  <div
    className="label-with-popover"
  >
    <span
      className="pf-c-label"
      onClick={[Function]}
      onPointerEnter={[Function]}
      onPointerLeave={[Function]}
    >
      <a
        className="pf-c-label__content"
        href="#"
      >
        <div
          className="pf-l-split pf-m-gutter channel-type-label"
        >
          <div
            className="pf-l-split__item"
          >
            Local cluster
          </div>
          <div
            className="pf-l-split__item"
          >
            <svg
              alt="Provide a description that will be used as the title"
              aria-label="Provide a description that will be used as the title"
              className="channel-type-icon"
              fillRule="evenodd"
              height="16"
              name="icon--launch"
              role="img"
              viewBox="0 0 16 16"
              width="16"
            >
              <title>
                Provide a description that will be used as the title
              </title>
              <path
                d="M14.3 1h-3.8V0H16v5.5h-1V1.7L9.7 7 9 6.3 14.3 1z"
              />
              <path
                d="M14.3 1h-3.8V0H16v5.5h-1V1.7L9.7 7 9 6.3 14.3 1z"
              />
              <path
                d="M13 9h1v6c0 .6-.4 1-1 1H1c-.6 0-1-.4-1-1V3c0-.6.4-1 1-1h7v1H1v12h12V9z"
              />
            </svg>
          </div>
        </div>
      </a>
    </span>
  </div>
</div>
`;
